### ISSUE TITLE
time incorrect

### ISSUE THREAD

**Issue by sunboy123** [kevinzg/facebook-scraper#218]
i find the time of posts is incorrect,  get time is greater than website time show

**Comment by neon-ninja** [comment]
Facebook doesn't provide exact times of posts if they're made recently, it says something like "3 hours ago", which is imprecise so leads to minor variation in interpreted time

**Comment by VariabileAleatoria** [comment]
so this is why the value of seconds for the same post is different every time I scrape?

**Comment by fashandatafields** [comment]
> so this is why the value of seconds for the same post is different every time I scrape?
> Example:
> 
> ```
> >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> >>> posts[0]['time']
> datetime.datetime(2021, 4, 24, 21, 3, 20, 245212)
> >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> >>> posts[0]['time']
> datetime.datetime(2021, 4, 24, 21, 2, 54, 563809)
> ```

So what u do, assign a variable to it and

**Comment by neon-ninja** [comment]
> so this is why the value of seconds for the same post is different every time I scrape?
> Example:
> 
> ```
> >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> >>> posts[0]['time']
> datetime.datetime(2021, 4, 24, 21, 3, 20, 245212)
> >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> >>> posts[0]['time']
> datetime.datetime(2021, 4, 24, 21, 2, 54, 563809)
> ```

Yes. Because of the incomplete time information, consider time values of recent posts just an approximation. If you need a unique id, use the post id.

**Comment by VariabileAleatoria** [comment]
> > so this is why the value of seconds for the same post is different every time I scrape?
> > Example:
> > ```
> > >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> > >>> posts[0]['time']
> > datetime.datetime(2021, 4, 24, 21, 3, 20, 245212)
> > >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> > >>> posts[0]['time']
> > datetime.datetime(2021, 4, 24, 21, 2, 54, 563809)
> > ```
> 
> Yes. Because of the incomplete time information, consider time values of recent posts just an approximation. If you need a unique id, use the post id.

yup that's exactly what I need, because I check periodically for new posts and I have to see if they were already considered.  
Do you know if ids are assigned to posts incrementally? (a later post in time will have  an higher id?)

**Comment by neon-ninja** [comment]
> > > so this is why the value of seconds for the same post is different every time I scrape?
> > > Example:
> > > ```
> > > >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> > > >>> posts[0]['time']
> > > datetime.datetime(2021, 4, 24, 21, 3, 20, 245212)
> > > >>> posts = list(get_posts(name_of_page, pages=2, cookies='cookies.txt'))
> > > >>> posts[0]['time']
> > > datetime.datetime(2021, 4, 24, 21, 2, 54, 563809)
> > > ```
> > 
> > 
> > Yes. Because of the incomplete time information, consider time values of recent posts just an approximation. If you need a unique id, use the post id.
> 
> yup that's exactly what I need, because I check periodically for new posts and I have to see if they were already considered.
> Do you know if ids are assigned to posts incrementally? (a later post in time will have an higher id?)

Yes, post ids appear to be incremental

**Comment by neon-ninja** [comment]
Something else I note - while m.facebook.com abbr elements are relatively bare:

```html
<abbr>16 March at 10:21</abbr>
```

unauthenticated w3 abbr elements are a lot more informative:

```html
<abbr data-utime="1594397700" title="Friday, 10 July 2020 at 09:15" data-shorten="1" class="_5ptz"><span class="timestampContent" id="js_1">10 July 2020</span></abbr>
```

Note the the data-utime attribute is a unix timestamp, which would be relatively easy/reliable to parse, and give accuracy to the second.

@kevinzg is it worth making an extra request just to get accurate time information?

**Comment by kevinzg** [comment]
@neon-ninja ideally the user would be able to tune it to their needs. For now we could decide looking at the `allow_extra_requests` option you added, and the w3 post page is already requested for some fields, so we could reuse the response.

**Comment by papaya-ais** [comment]
Maybe we can use publish_time in post['source'].
I'm trying to get this for exact time.

![image](https://user-images.githubusercontent.com/71307736/116679013-fac84a80-a9e4-11eb-8602-ee40741555b9.png)
![image](https://user-images.githubusercontent.com/71307736/116679068-07e53980-a9e5-11eb-8dc0-81e5f6c92f76.png)
![image](https://user-images.githubusercontent.com/71307736/116679290-52ff4c80-a9e5-11eb-8cfb-8494fc5393b3.png)
![image](https://user-images.githubusercontent.com/71307736/116679332-5d214b00-a9e5-11eb-9306-82019ede2a33.png)

**Comment by neon-ninja** [comment]
@papaya-ais can you clarify what you mean? I haven't seen anything called publish_time

**Comment by papaya-ais** [comment]
@neon-ninja 
Sorry for deleting images last comment but I had to remove that because my information was in image

If you add this line in extractors.py 134
`    self.extract_share_information,
            self.extract_availability,
        ]

        post = self.make_new_post()
        post['source'] = self.element
        print(post['source'])  # this

        # TODO: this is just used by `extract_reactions`, probably should not be acceded from self
        self.post = post
`

Than you can see this element 'article'

> <Element 'article' class=('_55wo', '_5rgr', '_5gh8', '_3drq', 'async_like') data-store='{"linkdata":"mf_story_k ...


And we can find publish time here...

> "post_context":{"object_fbtype":266,"publish_time":1619632211,"story_name":"EntStatusCreationStory","story_fbid":["123123123123123"]}

...and can get timestamp 1619632211

It can be transfer easily with datetime
`import datetime


if __name__ == '__main__':
    post_time = datetime.datetime.fromtimestamp(1619632211).strftime("%Y-%m-%d %H:%M:%S")
    print(post_time)
`
>2021-04-29 02:50:11

**Comment by neon-ninja** [comment]
I see - publish_time only shows up for anonymous requests, and not if you're logged in / passing cookies

**Comment by papaya-ais** [comment]
Oh I see. I only use this with anonymous request so I didn't know about that.
Thank you for your explanation.
